iOS Client/Store/MBArticlesStore.swift

```diff
@@ -106,49 +102,49 @@ class MBArticlesStore: NSObject {
+        return performDownload(clientFunction: client.getAuthorsWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBAuthor.deserialize)
     }
     
-    private func downloadCategories(managedObjectContext: NSManagedObjectContext, completion: @escaping (Bool?, Error?) -> Void) {
-        performDownload(clientFunction: client.getCategoriesWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBCategory.deserialize, completion: completion)
+    private func downloadCategories(managedObjectContext: NSManagedObjectContext) -> Promise<Bool> {
+        return performDownload(clientFunction: client.getCategoriesWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBCategory.deserialize)
     }
     
-    private func downloadArticles(managedObjectContext: NSManagedObjectContext, completion: @escaping (Bool?, Error?) -> Void) {
-        performDownload(clientFunction: client.getArticlesWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBArticle.deserialize, completion: completion)
+    private func downloadArticles(managedObjectContext: NSManagedObjectContext) -> Promise<Bool> {
+        return performDownload(clientFunction: client.getArticlesWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBArticle.deserialize)
     }
     
     // An internal helper function to perform a download
     private func performDownload(clientFunction: (@escaping ([Data], Error?) -> Void) -> Void,
                                  managedObjectContext: NSManagedObjectContext,
-                                 deserializeFunc: @escaping (NSDictionary, NSManagedObjectContext) throws -> Bool,
-                                 completion: @escaping (Bool?, Error?) -> Void) {
-        
-        clientFunction { (data: [Data], err: Error?) in
-            if let clientErr = err {
-                completion(false, clientErr)
-                return
-            }
-            
-            var isNewData = false
-            var caughtError: Error? = nil
-            // data is an array of Data, where each datum is a serialized array representing a page of results
-            // thus, think of data as an array of results pages
-            for jsonData in data {
-                do {
-                    isNewData = try self.downloadModelsHandler(managedObjectContext: managedObjectContext,
-                                                   data: jsonData,
-                                                   deserializeFunc: deserializeFunc) || isNewData
-                } catch {
-                    print("could not handle data: \(error) \(error.localizedDescription)")
-                    caughtError = error
+                                 deserializeFunc: @escaping (NSDictionary, NSManagedObjectContext) throws -> Bool) -> Promise<Bool> {
+        return Promise { fulfill, reject in
```

Comment:
might be able to do something with pending promises here. I'd have to take a closer look but I think we can make this cleaner. If it works its fine for now